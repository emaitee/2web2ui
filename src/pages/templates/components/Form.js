/* eslint max-lines: ["error", 200] */
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Field, change } from 'redux-form';
import { list as listDomains } from 'src/actions/sendingDomains';
import { selectVerifiedDomains } from 'src/selectors/sendingDomains';
import config from 'src/config';

// Components
import { Panel } from '@sparkpost/matchbox';
import ToggleBlock from './ToggleBlock';
import { TextFieldWrapper } from 'src/components';
import FromEmailWrapper from './FromEmail';

// Helpers & Validation
import { required } from 'src/helpers/validation';
import { slugify } from 'src/helpers/string';
import { ID_ALLOWED_CHARS, idSyntax, substitution, emailOrSubstitution } from './validation';

import styles from './FormEditor.module.scss';

export class Form extends Component {
  // Fills in ID based on Name
  handleIdFill = (e) => {
    const { newTemplate, change, name } = this.props;
    if (!newTemplate) {
      return;
    }

    const idValue = slugify(e.target.value).replace(new RegExp(`[^${ID_ALLOWED_CHARS}]`, 'g'), '');
    change(name, 'id', idValue);
  }

  componentDidMount() {
    this.props.listDomains();
  }

  componentWillReceiveProps(nextProps) {
    const { change, newTemplate, name } = this.props;
    const { domains } = nextProps;

    // If no verified sending domains, use sandbox
    if (newTemplate && !domains.length) {
      change(name, 'content.from.email', `${config.sandbox.localpart}@${config.sandbox.domain}`);
    }
  }

  verifiedDomain = (value) => {
    const { domains } = this.props;
    const parts = value.split('@');

    if (parts.length > 1) {
      const validSandbox = value === `${config.sandbox.localpart}@${config.sandbox.domain}`;
      const validDomain = parts[1] && (domains.map(({ domain }) => domain).includes(parts[1]) || !substitution(parts[1]));
      return validSandbox || validDomain ? undefined : 'Must use a verified sending domain';
    }

    return undefined;
  }

  render() {
    const { newTemplate, published, domains } = this.props;

    return (
      <Panel className={styles.FormPanel}>
        <Panel.Section>
          <Field
            name='name'
            component={TextFieldWrapper}
            label='Template Name'
            onChange={this.handleIdFill}
            disabled={published}
            validate={required}
          />

          <Field
            name='id'
            component={TextFieldWrapper}
            label='Template ID'
            helpText={'A Unique ID for your template, we\'ll fill this in for you.'}
            disabled={!newTemplate || published}
            validate={[required, idSyntax]}
          />
        </Panel.Section>

        <Panel.Section>
          <Field
            name='content.from.name'
            component={TextFieldWrapper}
            label='From Name'
            helpText='A friendly from for your recipients.'
            disabled={published}
          />

          <Field
            name='content.from.email'
            component={FromEmailWrapper}
            placeholder='example@email.com'
            label='From Email'
            disabled={!domains.length || published}
            validate={[required, emailOrSubstitution, this.verifiedDomain]}
            domains={domains}
          />

          <Field
            name='reply_to'
            component={TextFieldWrapper}
            label='Reply To'
            helpText='An email address recipients can reply to.'
            disabled={published}
            validate={emailOrSubstitution}
          />

          <Field
            name='content.subject'
            component={TextFieldWrapper}
            label='Subject'
            disabled={published}
            validate={required}
          />

          <Field
            name='description'
            component={TextFieldWrapper}
            label='Description'
            helpText='Not visible to recipients.'
            disabled={published}
          />
        </Panel.Section>

        <Panel.Section>
          <Field
            name='options.open_tracking'
            component={ToggleBlock}
            label='Track Opens'
            type='checkbox'
            parse={(value) => !!value} // Prevents unchecked value from equaling ""
            disabled={published}
          />

          <Field
            name='options.click_tracking'
            component={ToggleBlock}
            label='Track Clicks'
            type='checkbox'
            parse={(value) => !!value}
            disabled={published}
          />
        </Panel.Section>

        <Panel.Section>
          <Field
            name='options.transactional'
            component={ToggleBlock}
            label='Transactional'
            type='checkbox'
            parse={(value) => !!value}
            helpText='Transactional messages are triggered by a userâ€™s actions on the website, like requesting a password reset, signing up, or making a purchase.'
            disabled={published}
          />
        </Panel.Section>
      </Panel>
    );
  }
}

const mapStateToProps = (state) => ({
  domains: selectVerifiedDomains(state)
});

export default connect(mapStateToProps, { change, listDomains })(Form);
