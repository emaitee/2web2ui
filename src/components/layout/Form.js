import React from 'react';

import ScrollToTop from './ScrollToTop';
import { Loading } from 'src/components';
import styles from './Layout.module.scss';

const Form = ({ children, loading }) => (
  <div className={styles.wrapper}>
    <main role="main" className={styles.formContent}>
      <div className={styles.formContainer}>
        { loading ? <div className={styles.loading}><Loading /></div> : children }
      </div>
    </main>
    <ScrollToTop/>
  </div>
);

export default Form;
