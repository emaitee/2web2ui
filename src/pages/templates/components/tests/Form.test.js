import { shallow } from 'enzyme';
import React from 'react';

import { Form } from '../Form';

describe('Template Form', () => {
  let wrapper;

  beforeEach(() => {
    const props = {
      domains: [
        { domain: 'test.com' },
        { domain: 'verified.com' }
      ],
      listDomains: jest.fn(),
      change: jest.fn(),
      newTemplate: false,
      published: false,
      name: 'form-name'
    };

    wrapper = shallow(<Form {...props} />);
  });

  afterEach(() => {
    wrapper.unmount();
  });

  it('should render', () => {
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.instance().props.listDomains).toHaveBeenCalled();
  });

  it('should disable fields on published', () => {
    wrapper.setProps({ published: true });
    expect(wrapper).toMatchSnapshot();
  });

  it('should use sandbox domain if account has no verified domains', () => {
    wrapper.setProps({ newTemplate: true });
    wrapper.setProps({ domains: []});
    expect(wrapper.instance().props.change).toHaveBeenCalledWith(wrapper.instance().props.name, 'content.from.email', 'sandbox@sparkpostbox.com');
  });

  it('should not handle ID fill on edit', () => {
    wrapper.find('Field').at(0).simulate('change', { target: { value: 'test 1 2!' }});
    expect(wrapper.instance().props.change).not.toHaveBeenCalled();
  });

  it('should handle ID fill', () => {
    wrapper.setProps({ newTemplate: true });
    wrapper.find('Field').at(0).simulate('change', { target: { value: 'test 1 2!' }});
    expect(wrapper.instance().props.change).toHaveBeenCalledWith(wrapper.instance().props.name, 'id', 'test-1-2');
  });

  it('should validate verified domain', () => {
    wrapper.setProps({ domains: [{ domain: 'valid.com' }]});
    const result = wrapper.instance().verifiedDomain('email@valid.com');
    expect(result).toEqual(undefined);
  });

  it('should validate sandbox domain', () => {
    const result = wrapper.instance().verifiedDomain('sandbox@sparkpostbox.com');
    expect(result).toEqual(undefined);
  });

  it('should validate unverified domain', () => {
    wrapper.setProps({ domains: [{ domain: 'valid.com' }]});
    const result = wrapper.instance().verifiedDomain('email@notvalid.com');
    expect(result).toEqual('Must use a verified sending domain');
  });

  it('should validate substituted domain', () => {
    wrapper.setProps({ domains: [{ domain: 'valid.com' }]});
    const result = wrapper.instance().verifiedDomain('email@{{domain}}');
    expect(result).toEqual(undefined);
  });
});
