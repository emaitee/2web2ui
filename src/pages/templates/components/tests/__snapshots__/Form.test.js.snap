// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Template Form should disable fields on published 1`] = `
<Panel>
  <Panel.Section>
    <Field
      component={[Function]}
      disabled={true}
      label="Template Name"
      name="name"
      onChange={[Function]}
      validate={[Function]}
    />
    <Field
      component={[Function]}
      disabled={true}
      helpText="A Unique ID for your template, we'll fill this in for you."
      label="Template ID"
      name="id"
      validate={
        Array [
          [Function],
          [Function],
        ]
      }
    />
  </Panel.Section>
  <Panel.Section>
    <Field
      component={[Function]}
      disabled={true}
      helpText="A friendly from for your recipients."
      label="From Name"
      name="content.from.name"
    />
    <Field
      component={[Function]}
      disabled={true}
      domains={
        Array [
          Object {
            "domain": "test.com",
          },
          Object {
            "domain": "verified.com",
          },
        ]
      }
      label="From Email"
      name="content.from.email"
      placeholder="example@email.com"
      validate={
        Array [
          [Function],
          [Function],
          [Function],
        ]
      }
    />
    <Field
      component={[Function]}
      disabled={true}
      helpText="An email address recipients can reply to."
      label="Reply To"
      name="reply_to"
      validate={[Function]}
    />
    <Field
      component={[Function]}
      disabled={true}
      label="Subject"
      name="content.subject"
      validate={[Function]}
    />
    <Field
      component={[Function]}
      disabled={true}
      helpText="Not visible to recipients."
      label="Description"
      name="description"
    />
  </Panel.Section>
  <Panel.Section>
    <Field
      component={[Function]}
      disabled={true}
      label="Track Opens"
      name="options.open_tracking"
      parse={[Function]}
      type="checkbox"
    />
    <Field
      component={[Function]}
      disabled={true}
      label="Track Clicks"
      name="options.click_tracking"
      parse={[Function]}
      type="checkbox"
    />
  </Panel.Section>
  <Panel.Section>
    <Field
      component={[Function]}
      disabled={true}
      helpText="Transactional messages are triggered by a user’s actions on the website, like requesting a password reset, signing up, or making a purchase."
      label="Transactional"
      name="options.transactional"
      parse={[Function]}
      type="checkbox"
    />
  </Panel.Section>
</Panel>
`;

exports[`Template Form should render 1`] = `
<Panel>
  <Panel.Section>
    <Field
      component={[Function]}
      disabled={false}
      label="Template Name"
      name="name"
      onChange={[Function]}
      validate={[Function]}
    />
    <Field
      component={[Function]}
      disabled={true}
      helpText="A Unique ID for your template, we'll fill this in for you."
      label="Template ID"
      name="id"
      validate={
        Array [
          [Function],
          [Function],
        ]
      }
    />
  </Panel.Section>
  <Panel.Section>
    <Field
      component={[Function]}
      disabled={false}
      helpText="A friendly from for your recipients."
      label="From Name"
      name="content.from.name"
    />
    <Field
      component={[Function]}
      disabled={false}
      domains={
        Array [
          Object {
            "domain": "test.com",
          },
          Object {
            "domain": "verified.com",
          },
        ]
      }
      label="From Email"
      name="content.from.email"
      placeholder="example@email.com"
      validate={
        Array [
          [Function],
          [Function],
          [Function],
        ]
      }
    />
    <Field
      component={[Function]}
      disabled={false}
      helpText="An email address recipients can reply to."
      label="Reply To"
      name="reply_to"
      validate={[Function]}
    />
    <Field
      component={[Function]}
      disabled={false}
      label="Subject"
      name="content.subject"
      validate={[Function]}
    />
    <Field
      component={[Function]}
      disabled={false}
      helpText="Not visible to recipients."
      label="Description"
      name="description"
    />
  </Panel.Section>
  <Panel.Section>
    <Field
      component={[Function]}
      disabled={false}
      label="Track Opens"
      name="options.open_tracking"
      parse={[Function]}
      type="checkbox"
    />
    <Field
      component={[Function]}
      disabled={false}
      label="Track Clicks"
      name="options.click_tracking"
      parse={[Function]}
      type="checkbox"
    />
  </Panel.Section>
  <Panel.Section>
    <Field
      component={[Function]}
      disabled={false}
      helpText="Transactional messages are triggered by a user’s actions on the website, like requesting a password reset, signing up, or making a purchase."
      label="Transactional"
      name="options.transactional"
      parse={[Function]}
      type="checkbox"
    />
  </Panel.Section>
</Panel>
`;
